Prevent non-video file formats

mushtaqak opened the pull request:
# Description

This PR adds the ability to prevent non-video file uploads on Video Uploads page. 

### Supported file formats
Currently, supported file formats are .mp4 (video/mp4) and .mov (video/quicktime). For more info, [read the docs](http://edx.readthedocs.io/projects/edx-partner-course-staff/en/latest/video/video_uploads.html#supported-video-formats).


### [TNL-5956](https://openedx.atlassian.net/browse/TNL-5956)

### Sandbox
https://studio-preventnonvideo.sandbox.edx.org/videos/course-v1:edX+DemoX+Demo_Course
### Testing
- [x] Unit
- [x] Jasmine
- [x] Manual
### Reviewers

If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @muzaffaryousaf 
- [x] Code review: @muhammad-ammar 
- [x] Doc: @catong  
- [x] a11y: @cptvitamin   
- [x] Product: @sstack22 
### Post-review
- [x] Squash commits

muzaffaryousaf commented on cms/djangoapps/contentstore/views/videos.py:
```diff
@@ -297,6 +297,7 @@ def videos_post(course, request):
 
     The returned array corresponds exactly to the input array.
     """
+    supported_content_types = ['video/mp4', 'video/quicktime']
```
@mushtaqak Let's restrict only non-video files for now.

mushtaqak replied:
There are so many non-video files. I think, the best is we check for supported file formats.

mushtaqak commented:
jenkins run js

mushtaqak commented:
jenkins run js

muhammad-ammar commented on cms/static/js/views/active_video_upload_list.js:
```diff
@@ -1,8 +1,19 @@
+             activeVideoUploadListTemplate) {
         'use strict';
-
+        var supportedFileFormats = ['mp4', 'mov'];
```
@mushtaqak We should have a list or dict of supported filetypes at the backend and then we can use it here.

muhammad-ammar commented on cms/static/js/views/active_video_upload_list.js:
```diff
@@ -162,13 +184,43 @@ define(
+                    fileName = file.name;
+                    fileType = fileName.substr(fileName.lastIndexOf('.') + 1);
+                    // validate file type
+                    if (supportedFileFormats.indexOf(fileType) === -1) {
```
we can use [_.contains](http://underscorejs.org/#contains) here instead of `indexof`

muhammad-ammar commented on cms/static/js/views/active_video_upload_list.js:
```diff
@@ -162,13 +184,43 @@ define(
+                    if (supportedFileFormats.indexOf(fileType) === -1) {
+                        error = gettext('`{filename}` has unsupported file format.')
+                                        .replace('{filename}', fileName);
+                        return false;
```
Can we show message here and return from here? If yes then we can avoid the `if(error)` below.

muhammad-ammar commented on cms/static/js/views/active_video_upload_list.js:
```diff
@@ -162,13 +184,43 @@ define(
                 if (this.onFileUploadDone) {
                     this.onFileUploadDone(this.collection);
                     this.clearSuccessful();
+                    this.fileErrorMsg = null;
```
Why we need this? We should hide error message before a new upload.

muhammad-ammar commented on cms/static/js/views/active_video_upload_list.js:
```diff
@@ -22,10 +33,15 @@ define(
                     options.uploadButton.click(this.chooseFile.bind(this));
                 }
+                // error message modal for file uploads
+                this.fileErrorMsg = null;
```
we are setting this to `null` at two places. I wonder if we really need this variable?

muhammad-ammar commented on cms/static/js/spec/views/active_video_upload_list_spec.js:
```diff
@@ -59,6 +59,31 @@ define(
+                        ]
+                    };
+                    this.view.$uploadForm.fileupload('add', supportedFiles);
+                    expect(this.view.fileErrorMsg).toBeNull();
```
Instead of checking `fileErrorMsg` for `null`, i think it would be better to explicitly check the notificationview visibility. I think we should remove `fileErrorMsg` from view as i mentioned earlier.

muhammad-ammar commented on cms/static/js/views/active_video_upload_list.js:
```diff
@@ -162,13 +184,43 @@ define(
+                    fileType = fileName.substr(fileName.lastIndexOf('.') + 1);
+                    // validate file type
+                    if (supportedFileFormats.indexOf(fileType) === -1) {
+                        error = gettext('`{filename}` has unsupported file format.')
```
Should we also show in the error message what are supported file formats?
@muzaffaryousaf Any thoughts?

muhammad-ammar commented on cms/static/js/views/active_video_upload_list.js:
```diff
@@ -162,13 +184,43 @@ define(
+                    title: gettext('Your file could not be uploaded'),
+                    message: error
+                });
+                this.fileErrorMsg.show();
```
Instead of the setting the `fileErrorMsg` variable on view use a function level variable to show the message and then in tests verify if the error message is present in `DOM` or not. This should work.

mushtaqak replied:
We want this because we want this notification to disable too.

muhammad-ammar commented on cms/static/js/spec/views/active_video_upload_list_spec.js:
```diff
@@ -59,6 +59,31 @@ define(
+                    this.view.$uploadForm.fileupload('add', unSupportedFiles);
+                    expect(this.view.fileErrorMsg).toBeDefined();
+                    expect(this.view.fileErrorMsg.options.title).toEqual('Your file could not be uploaded');
+                    expect(this.view.fileErrorMsg.options.message).toEqual('`test-3.txt` has unsupported file format.');
```
Same comment as above.

muhammad-ammar commented on cms/djangoapps/contentstore/views/tests/test_videos.py:
```diff
@@ -223,15 +223,80 @@ def assert_bad(content):
+        self.assertEqual(response.status_code, 400)
+        response = json.loads(response.content)
+        self.assertIn('error', response)
+        self.assertEqual(response['error'], "Request 'files' entry contain unsupported content_type")
```
What about having a loop for both supported and unsupported like below? 

```Python
for files, status in ((supported_files, 200), (unsupported_files, 400)):
    response = self.client.post(
        self.url,
        json.dumps({"files": files}),
        content_type="application/json"
    )

    self.assertEqual(response.status_code, status)
    response = json.loads(response.content)
    
    if status == 200:
        self.assertNotIn('error', response)
    else:
        self.assertIn('error', response)
        self.assertEqual(response['error'], "Request 'files' entry contain unsupported content_type")
```

mushtaqak replied:
makes sense, I would probably achieve this with jelp of `ddt` as well.

muhammad-ammar commented on cms/djangoapps/contentstore/views/tests/test_videos.py:
```diff
@@ -223,15 +223,80 @@ def assert_bad(content):
+            for file_info in supported_files + unsupported_files
+        ]
+        # If extra calls are made, return a dummy
+        mock_key.side_effect = mock_key_instances + [Mock()]
```
What exactly we are doing here?

mushtaqak replied:
We are mocking here to upload videos.

muzaffaryousaf commented:
@muhammad-ammar You can complete your review today.
@mushtaqak Let me know when you're done with addressing ammar's feedback.

mushtaqak commented:
@muhammad-ammar addressed your feedbacks.

muzaffaryousaf commented on cms/djangoapps/contentstore/views/videos.py:
```diff
@@ -305,6 +311,11 @@ def videos_post(course, request):
         error = "Request 'files' entry does not contain 'file_name' and 'content_type'"
+    elif any(
+            file['content_type'] not in VIDEO_SUPPORTED_FILE_FORMATS.values()
+            for file in request.json["files"]
```
Can we use `request.json.get('files', [])` instead ?

muzaffaryousaf replied:
Write a test which verifies the behavior when we empty list.

mushtaqak replied:
I think, we don't need `request.json.get('files', []) ` because if `files` key is not present, [first condition](https://github.com/edx/edx-platform/blob/master/cms/djangoapps/contentstore/views/videos.py#L301) of the if block would be true and this code will never run.

Also, there is already a [test](https://github.com/edx/edx-platform/blob/master/cms/djangoapps/contentstore/views/tests/test_videos.py#L197) written to check empty list and this behaviour.

muzaffaryousaf commented on cms/djangoapps/contentstore/views/videos.py:
```diff
@@ -305,6 +311,11 @@ def videos_post(course, request):
+            file['content_type'] not in VIDEO_SUPPORTED_FILE_FORMATS.values()
+            for file in request.json["files"]
+    ):
+        error = "Request 'files' entry contain unsupported content_type"
```
@catong Can you review doc strings in this PR.

muzaffaryousaf replied:
@mushtaqak I don't think this is user friendly message.

mushtaqak replied:
@muzaffaryousaf I have made this because messages here backend are like that, and normal user would not get this, unless he sends some ajax through console or any other way. User facing message comes from JS-side.

muzaffaryousaf commented:
@sstack22 Please give it a quick look. Thanks.

mushtaqak commented on cms/static/js/views/active_video_upload_list.js:
```diff
@@ -169,6 +196,43 @@ define(
+                    // validate file type
+                    if (!_.contains(self.videoSupportedFileFormats, fileType)) {
+                        error = gettext(
+                            '`{filename}` has unsupported file format. ' +
```
@catong Can you review error message here. This is the message that is shown to user.

catong replied:
```{filename} is not in a supported file format.```

catong replied:
```{filename} is not in a supported file format.```

muzaffaryousaf commented:
@mushtaqak Make sure that you run safe-commit job locally. Also, don't forget to verify xss related changes manually to make sure that things are working as expected.

mushtaqak commented on cms/static/js/views/active_video_upload_list.js:
```diff
@@ -169,6 +196,43 @@ define(
+                        )
+                        .replace('{filename}', fileName)
+                        .replace('{supportedFileFormats}', self.videoSupportedFileFormats.join(', '));
+                        return false;
```
Yes this does

muzaffaryousaf replied:
Well i don't think so. Can you double check ?

muzaffaryousaf replied:
Here is a note from doc ` It's also good to note that an each loop cannot be broken out of — to break, use _.find instead.`

muzaffaryousaf replied:
I see, you're using each from JQuery

mushtaqak replied:
@muzaffaryousaf I think, we can't use `_.find` here because, we just don't need to break the loop on simple condition we need to set error message too.

Either we use `$.each` or traditional `for(int i=0;i<array.length;i++) {`  javascript loop.

sstack22 commented:
@mushtaqak , @muzaffaryousaf  - is there a sandbox I can review?

muzaffaryousaf commented:
@mushtaqak Were you able to build a sandbox successfully ?

mushtaqak commented on cms/static/js/spec/views/active_video_upload_list_spec.js:
```diff
@@ -59,6 +61,34 @@ define(
+                    expect(this.view.fileErrorMsg.options.message).toEqual(
+                        'test-3.txt is not in a supported file format. Supported file formats are "' +
+                        this.videoSupportedFileFormats.join(', ') + '".'
+                    );
```
@catong The complete mesage shown to user would be like :

> Your file could not be uploaded
> Read More.txt is not in a supported file format. Supported file formats are "mp4, mov".

I am creating a sandbox, hopefully I will be able to build this time.

catong replied:
@mushtaqak Thanks for the sandbox! Can we make a small change to the punctuation in the message? In the 2nd sentence, can we remove the double quotation marks around the supported file extensions, add the periods in front of each extension, and use "and" between the 2 extensions: "Supported file formats are .mp4 and .mov."

mushtaqak commented:
@sstack22 We are currently blocked on sandbox creation. I will let you know when sandbox is created.

mushtaqak commented:
@muzaffaryousaf @sstack22 Sandbox built https://studio-nonvideo.sandbox.edx.org/videos/course-v1:edX+DemoX+Demo_Course

@catong FYI on error message when you upload a txt or png file (non-video).

muzaffaryousaf commented:
@mushtaqak Once i upload not supported file, i kept seeing error notification for supported files too. Can you look into that.

muzaffaryousaf commented on cms/static/js/views/active_video_upload_list.js:
```diff
@@ -101,7 +117,18 @@ define(
+                errorMsg = view.validateFile(uploadData);
+                if (errorMsg) {
+                    view.showErrorMessage(errorMsg);
+                    return;
```
We can avoid this return if we move rest of the code in this function in else part.

muzaffaryousaf commented:
@mushtaqak I'm done with first pass, let me know when it's ready for 2nd review.

mushtaqak commented:
@muzaffaryousaf Addressed feedback.

Building a new sandbox : https://studio-nonvideoformat.sandbox.edx.org/videos/course-v1:edX+DemoX+Demo_Course

catong commented on cms/templates/videos_index.html:
```diff
@@ -66,10 +72,19 @@ <h3 class="title-3">${_("Maximum Video File Size")}</h3>
+                )}</p>
                 <h3 class="title-3">${_("How do I get the videos into my course?")}</h3>
-                <p>${_("When status for a file is {em_start}Ready{em_end}, you can add that video to a component in your course. Copy the unique video ID. In another browser window, on the Course Outline page, create or locate a video component to play this video. Edit the video component to paste the ID into the Advanced {em_start}Video ID{em_end} field. The video can play in the LMS as soon as  ...
+                <p>${Text(_("When status for a file is {em_start}Ready{em_end}, you can add that video to a component in your course. Copy the unique video ID. In another browser window, on the Course Outline page, create or locate a video component to play this video. Edit the video component to paste the ID into the Advanced {em_start}Video ID{em_end} field. The video can play in the LMS as soo ...
```
@mushtaqak Please add missing word "the": "When ^the^ status for a file..."

catong commented:
👍 from me with punctuation changes to the error message, and small tweak to the sidebar text. Thanks!

muzaffaryousaf commented on cms/static/js/views/active_video_upload_list.js:
```diff
@@ -169,6 +198,44 @@ define(
+            hideErrorMessage: function() {
+                if (this.fileErrorMsg) {
+                    this.fileErrorMsg.hide();
+                    this.fileErrorMsg = null;
```
@mushtaqak Any reason for making it null and hide ?

muzaffaryousaf replied:
Make sure that we're following same approach like assets page if applicable.

mushtaqak replied:
Asset page page hides the notfication view when modal is closed (they have custom modal), and they make the errorMsg variable null when the process is done. 

We are doing the same but we have a different flow. We are hiding the notification view and making it null when file-chooser click event is called so that it won' come here again (Remember notification.shownAt property would be invalidated once hide is called, so next time, hide would render), that's why I am making this null and so does asset page at different locations.

muzaffaryousaf commented:
@mushtaqak This LGTM except few questions. 👍

muzaffaryousaf commented:
@mushtaqak This LGTM except few questions. 👍

mushtaqak commented:
jenkins run bokchoy

mushtaqak commented:
@sstack22 Waiting on your review. Here is the [sandbox video uploads page](https://studio-preventnonvideo.sandbox.edx.org/videos/course-v1:edX+DemoX+Demo_Course). 

I will merge when I get your thumbs-up.

catong commented:
👍 These changes are great! Thanks!

sstack22 commented:
👍 on my end. Has this error message been reviewed for a11y? As long as the pattern is accessible, we should be fine, but wanted to confirm we've looked into it.

mushtaqak commented:
@cptvitamin can you do a11y review ?

cptvitamin commented:
@mushtaqak The message that informs the user that the file was not uploaded is not being read to screen readers.  Unfortunately, putting an aria-live on the page-notifications region doesn't work because of the way that page gets refreshed.  Other teams have had success updating the #reader-feadback div that is on that page using the [accessibility_tools](https://github.com/edx/edx-platform/blob/master/common/static/js/src/accessibility_tools.js) methods.  See:
https://github.com/edx/edx-platform/pull/12633/files#diff-b09151dc6c39b2687a894e20cc943776R197

mushtaqak commented:
jenkins rull all

mushtaqak commented:
jenkins run all

mushtaqak commented:
@cptvitamin Thanks for the input. I have addressed your concerns. Kindly check again. [Sandbox](https://studio-preventnonvideo.sandbox.edx.org/videos/course-v1:edX+DemoX+Demo_Course) is updated :)

mushtaqak commented:
jenkins run all

mushtaqak commented:
jenkins run all

mushtaqak commented:
jenkins run bokchoy

mushtaqak commented:
@cptvitamin Can you review this today?

cptvitamin commented:
👍

cpennington commented:
This PR is being deployed to edx.org Staging in preparation for a release to production on Monday, 12/19 at 10 am. If you feel you need to validate it manually, please do so by then, and contact me if there is an urgent enough issue to scrap the deployment.

cpennington commented:
This PR is being deployed to edx.org Staging in preparation for a release to production on Tuesday, 12/20 at 10 am. If you feel you need to validate it manually, please do so by then, and contact me if there is an urgent enough issue to scrap the deployment.

mushtaqak commented:
@cpennington I have manually validated on the Staging. This is good to go :)

cpennington commented:
This PR has been deployed to edx.org Staging in preparation for a release to production on Wednesday, 12/21/16 at 10 am. If you feel you need to validate it manually, please do so by then, and contact me if there is an urgent enough issue to scrap the deployment.
